exchange/src/exchange/process_spawning/spawning.cpp

```diff
@@ -17,29 +17,30 @@ quote_id(std::string user_id)
 }
 
 pid_t
-spawn_client(auto& trader, const std::string& binary_path)
+spawn_client(
+    const std::unique_ptr<manager::generic_trader_t>& trader,
+    const std::string& binary_path
+)
 {
     using trader_type = std::decay_t<decltype(trader)>;
     if constexpr (std::is_same_v<trader_type, nutc::manager::local_trader_t>) {
-        const std::string filepath = trader.get_algo_path() + ".py";
+        const std::string filepath = trader->get_algo_id() + ".py";
         assert(file_ops::file_exists(filepath));
     }
```

Comment:
first of all this doesn't work anymore

second of all this shouldn't even be in here. its not proper seperation of concerns. you should not be worrying about the types of objects you get. you should only use the interface provided by `GenericTrader`. And if the interface doesn't support everything you need, then add to it.